File: go/101.go
/*

https://leetcode.com/problems/symmetric-tree/

Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

For example, this binary tree [1,2,2,3,4,4,3] is symmetric:

    1
   / \
  2   2
 / \ / \
3  4 4  3


But the following [1,2,2,null,3,null,3] is not:

    1
   / \
  2   2
   \   \
   3    3

*/

package main


import "fmt"


/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}


func isSymmetric(root *TreeNode) bool {
    var f func(a, b *TreeNode) bool 
    f = func(a, b *TreeNode) bool {
        if a == nil && b == nil {
            return true
        }
        if a == nil && b != nil || b == nil && a != nil {
            return false
        }
        if a.Val != b.Val {
            return false
        }
        return f(a.Left, b.Right) && f(a.Right, b.Left)
    }
    if root == nil {
        return true
    }
    return f(root.Left, root.Right)
}


func main() {

}

File: python/101.py
"""
https://leetcode.com/problems/symmetric-tree/

Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

For example, this binary tree [1,2,2,3,4,4,3] is symmetric:

    1
   / \
  2   2
 / \ / \
3  4 4  3
 

But the following [1,2,2,null,3,null,3] is not:

    1
   / \
  2   2
   \   \
   3    3

"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:

        def f(a, b):
            if not a and not b:
                return True
            if not a or not b:
                return False
            if a.val != b.val:
                return False
            return f(a.left, b.right) and f(a.right, b.left)

        return f(root.left, root.right) if root else True

# 没搞清楚为什么过了
# 反正就是过了

